File: artifacts/tulipindicators/decompose/src/fisher.c
int ti_fisher_start(const double *options)
{
  return ((int) options[0]) - 1;
}


/*** DEPENDENCIES:
***/


// hint:  ['signal_idx_ref is a mutable refrence to unsigned int', 'maxi_ref is a mutable refrence to int', 'mini_ref is a mutable refrence to int', 'max_ref is a mutable refrence to double', 'min_ref is a mutable refrence to double', 'val1_ref is a mutable refrence to double', 'bar_ref is a mutable refrence to double', 'fish_ref is a mutable refrence to double', 'j_ref is a mutable refrence to int']
void helper_ti_fisher_1(unsigned int * const signal_idx_ref, int * const maxi_ref, int * const mini_ref, double * const max_ref, double * const min_ref, double * const val1_ref, double * const bar_ref, double * const fish_ref, int * const j_ref, const double * const high, const double * const low, double * const fisher, double * const signal, int trail, int i)
{
  unsigned int signal_idx = *signal_idx_ref;
  int maxi = *maxi_ref;
  int mini = *mini_ref;
  double max = *max_ref;
  double min = *min_ref;
  double val1 = *val1_ref;
  double bar = *bar_ref;
  double fish = *fish_ref;
  int j = *j_ref;
  bar = 0.5 * (high[i] + low[i]);
  if (maxi < trail)
  {
    maxi = trail;
    max = 0.5 * (high[maxi] + low[maxi]);
    j = trail;
    while ((++j) <= i)
    {
      bar = 0.5 * (high[j] + low[j]);
      if (bar >= max)
      {
        max = bar;
        maxi = j;
      }
    }

  }
  else
    if (bar >= max)
  {
    maxi = i;
    max = bar;
  }
  bar = 0.5 * (high[i] + low[i]);
  if (mini < trail)
  {
    mini = trail;
    min = 0.5 * (high[mini] + low[mini]);
    j = trail;
    while ((++j) <= i)
    {
      bar = 0.5 * (high[j] + low[j]);
      if (bar <= min)
      {
        min = bar;
        mini = j;
      }
    }

  }
  else
    if (bar <= min)
  {
    mini = i;
    min = bar;
  }
  double mm = max - min;
  if (mm == 0.0)
  {
    mm = 0.001;
  }
  val1 = ((0.33 * 2.0) * ((((0.5 * (high[i] + low[i])) - min) / mm) - 0.5)) + (0.67 * val1);
  if (val1 > 0.99)
  {
    val1 = .999;
  }
  if (val1 < (-0.99))
  {
    val1 = -.999;
  }
  signal[signal_idx] = fish;
  signal_idx += 1;
  fish = (0.5 * log((1.0 + val1) / (1.0 - val1))) + (0.5 * fish);
  *(fisher++) = fish;
  *signal_idx_ref = signal_idx;
  *maxi_ref = maxi;
  *mini_ref = mini;
  *max_ref = max;
  *min_ref = min;
  *val1_ref = val1;
  *bar_ref = bar;
  *fish_ref = fish;
  *j_ref = j;
}


/*** DEPENDENCIES:
***/


int ti_fisher(int size, const double * const *inputs, const double *options, double * const *outputs)
{
  const double *high = inputs[0];
  const double *low = inputs[1];
  double *fisher = outputs[0];
  double *signal = outputs[1];
  unsigned int signal_idx = 0;
  const int period = (int) options[0];
  if (period < 1)
  {
    return 1;
  }
  if (size <= ti_fisher_start(options))
  {
    return 0;
  }
  int trail = 0;
  int maxi = -1;
  int mini = -1;
  double max = 0.5 * (high[0] + low[0]);
  double min = 0.5 * (high[0] + low[0]);
  double val1 = 0.0;
  double bar;
  double fish = 0.0;
  int i;
  int j;
  for (i = period - 1; i < size; i += 1, trail += 1)
  {
    helper_ti_fisher_1(&signal_idx, &maxi, &mini, &max, &min, &val1, &bar, &fish, &j, high, low, fisher, signal, trail, i);
  }

  assert((fisher - outputs[0]) == (size - ti_fisher_start(options)));
  assert(((&signal[signal_idx]) - outputs[1]) == (size - ti_fisher_start(options)));
  return 0;
}


/*** DEPENDENCIES:
int ti_fisher_start(const double *options)
{
  return ((int) options[0]) - 1;
}


----------------------------
void helper_ti_fisher_1(unsigned int * const signal_idx_ref, int * const maxi_ref, int * const mini_ref, double * const max_ref, double * const min_ref, double * const val1_ref, double * const bar_ref, double * const fish_ref, int * const j_ref, const double * const high, const double * const low, double * const fisher, double * const signal, int trail, int i)
{
  unsigned int signal_idx = *signal_idx_ref;
  int maxi = *maxi_ref;
  int mini = *mini_ref;
  double max = *max_ref;
  double min = *min_ref;
  double val1 = *val1_ref;
  double bar = *bar_ref;
  double fish = *fish_ref;
  int j = *j_ref;
  bar = 0.5 * (high[i] + low[i]);
  if (maxi < trail)
  {
    maxi = trail;
    max = 0.5 * (high[maxi] + low[maxi]);
    j = trail;
    while ((++j) <= i)
    {
      bar = 0.5 * (high[j] + low[j]);
      if (bar >= max)
      {
        max = bar;
        maxi = j;
      }
    }

  }
  else
    if (bar >= max)
  {
    maxi = i;
    max = bar;
  }
  bar = 0.5 * (high[i] + low[i]);
  if (mini < trail)
  {
    mini = trail;
    min = 0.5 * (high[mini] + low[mini]);
    j = trail;
    while ((++j) <= i)
    {
      bar = 0.5 * (high[j] + low[j]);
      if (bar <= min)
      {
        min = bar;
        mini = j;
      }
    }

  }
  else
    if (bar <= min)
  {
    mini = i;
    min = bar;
  }
  double mm = max - min;
  if (mm == 0.0)
  {
    mm = 0.001;
  }
  val1 = ((0.33 * 2.0) * ((((0.5 * (high[i] + low[i])) - min) / mm) - 0.5)) + (0.67 * val1);
  if (val1 > 0.99)
  {
    val1 = .999;
  }
  if (val1 < (-0.99))
  {
    val1 = -.999;
  }
  signal[signal_idx] = fish;
  signal_idx += 1;
  fish = (0.5 * log((1.0 + val1) / (1.0 - val1))) + (0.5 * fish);
  *(fisher++) = fish;
  *signal_idx_ref = signal_idx;
  *maxi_ref = maxi;
  *mini_ref = mini;
  *max_ref = max;
  *min_ref = min;
  *val1_ref = val1;
  *bar_ref = bar;
  *fish_ref = fish;
  *j_ref = j;
}


----------------------------
***/

File: artifacts/tulipindicators/translate/src/fisher.rs
use crate::*;
use std::f64::consts::LN_2;
use std::assert;

pub fn ti_fisher_start(options: Option<&[f64]>) -> Option<i32> {
    options.map(|opts| opts[0] as i32 - 1)
}

pub fn helper_ti_fisher_1(
    fisher_idx_ref: &mut u32,
    signal_idx_ref: &mut u32,
    maxi_ref: &mut i32,
    mini_ref: &mut i32,
    max_ref: &mut f64,
    min_ref: &mut f64,
    val1_ref: &mut f64,
    bar_ref: &mut f64,
    fish_ref: &mut f64,
    j_ref: &mut i32,
    high: &[f64],
    low: &[f64],
    fisher: &mut [f64],
    signal: &mut [f64],
    trail: i32,
    i: i32,
) {
    let mut fisher_idx = *fisher_idx_ref;
    let mut signal_idx = *signal_idx_ref;
    let mut maxi = *maxi_ref;
    let mut mini = *mini_ref;
    let mut max = *max_ref;
    let mut min = *min_ref;
    let mut val1 = *val1_ref;
    let mut bar = *bar_ref;
    let mut fish = *fish_ref;
    let mut j = *j_ref;

    bar = 0.5 * (high[i as usize] + low[i as usize]);
    
    if maxi < trail {
        maxi = trail;
        max = 0.5 * (high[maxi as usize] + low[maxi as usize]);
        j = trail;
        while {
            j += 1;
            j <= i
        } {
            bar = 0.5 * (high[j as usize] + low[j as usize]);
            if bar >= max {
                max = bar;
                maxi = j;
            }
        }
    } else if bar >= max {
        maxi = i;
        max = bar;
    }

    bar = 0.5 * (high[i as usize] + low[i as usize]);
    
    if mini < trail {
        mini = trail;
        min = 0.5 * (high[mini as usize] + low[mini as usize]);
        j = trail;
        while {
            j += 1;
            j <= i
        } {
            bar = 0.5 * (high[j as usize] + low[j as usize]);
            if bar <= min {
                min = bar;
                mini = j;
            }
        }
    } else if bar <= min {
        mini = i;
        min = bar;
    }

    let mut mm = max - min;
    if mm == 0.0 {
        mm = 0.001;
    }

    val1 = ((0.33 * 2.0) * ((((0.5 * (high[i as usize] + low[i as usize])) - min) / mm) - 0.5)) + (0.67 * val1);
    if val1 > 0.99 {
        val1 = 0.999;
    }
    if val1 < -0.99 {
        val1 = -0.999;
    }

    signal[signal_idx as usize] = fish;
    signal_idx += 1;
    fish = (0.5 * ((1.0 + val1) / (1.0 - val1)).ln()) + (0.5 * fish);
    fisher[fisher_idx as usize] = fish;
    fisher_idx += 1;

    *fisher_idx_ref = fisher_idx;
    *signal_idx_ref = signal_idx;
    *maxi_ref = maxi;
    *mini_ref = mini;
    *max_ref = max;
    *min_ref = min;
    *val1_ref = val1;
    *bar_ref = bar;
    *fish_ref = fish;
    *j_ref = j;
}
pub fn ti_fisher(
    size: i32,
    inputs: Option<&[&[f64]]>,
    options: Option<&[f64]>,
    outputs: Option<&mut [&mut [f64]]>,
) -> Option<i32> {
    // Check for None inputs/outputs
    let inputs = inputs?;
    let options = options?;
    let outputs = outputs?;

    let high = inputs[0];
    let low = inputs[1];
    
    // Split outputs into two mutable parts
    let (fisher_slice, signal_slice) = outputs.split_at_mut(1);
    let fisher = &mut fisher_slice[0];
    let signal = &mut signal_slice[0];

    let period = options[0] as i32;
    if period < 1 {
        return Some(1);
    }

    let start = ti_fisher_start(Some(options))?;
    if size <= start {
        return Some(0);
    }

    let mut trail = 0;
    let mut maxi = -1;
    let mut mini = -1;
    let mut max = 0.5 * (high[0] + low[0]);
    let mut min = 0.5 * (high[0] + low[0]);
    let mut val1 = 0.0;
    let mut bar = 0.0;
    let mut fish = 0.0;
    let mut j = 0;

    let mut fisher_idx = 0;
    let mut signal_idx = 0;

    for i in period - 1..size {
        helper_ti_fisher_1(
            &mut fisher_idx,
            &mut signal_idx,
            &mut maxi,
            &mut mini,
            &mut max,
            &mut min,
            &mut val1,
            &mut bar,
            &mut fish,
            &mut j,
            high,
            low,
            fisher,
            signal,
            trail,
            i,
        );
        trail += 1;
    }

    assert!((fisher_idx as i32) == (size - start));
    assert!((signal_idx as i32) == (size - start));

    Some(0)
}
